src/App.tsx

```diff
@@ -1,61 +1,147 @@
+import React, { useState } from 'react';
 import './App.scss';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import { TodoList } from './components/TodoList';
+import { Todo } from './types/Todo';
+
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+
+const titlePattern = /^[a-zA-Zа-яА-ЯїЇіІ0-9\s]*$/; //pattern that matches 'ua' and 'en' letters, digits and spaces
+
+function getUserByUserId(userId: number) {
+  return usersFromServer.find(user => user.id === userId) || null;
+}
+
+function getPreparedTodos() {
+  const preparedTodos = todosFromServer.map(todo => ({
+    ...todo,
+    user: getUserByUserId(todo.userId),
+  }));
+
+  return preparedTodos;
+}
+
+export const App: React.FC = () => {
+  const todos = getPreparedTodos();
+
+  const [visibleTodos, setVisibleTodos] = useState(todos);
```

Comment:
You can prepare todos once outside the component and pass them to the state directly